nixos/roles/devhost/vm.nix

```diff
@@ -0,0 +1,269 @@
+      case $opt in
+        m) memory="$OPTARG"
+        ;;
+        c) cores="$OPTARG"
+        ;;
+        \?) echo "Invalid option -$OPTARG" >&2
+        exit 1
+        ;;
+      esac
+    done
+    shift $((OPTIND-1))
+
+    action=''${1?need to specify action}
+
+    manage_alias_proxy=${lib.boolToString cfg.enableAliasProxy}
+
+    mkdir -p /etc/devhost/vm-configs
+    if [[ ! -f "/var/lib/devhost/ssh_bootstrap_key" ]]; then
+       cat > /var/lib/devhost/ssh_bootstrap_key <<EOF
+-----BEGIN OPENSSH PRIVATE KEY-----
+b3BlbnNzaC1rZXktdjEAAAAABG5vbmUAAAAEbm9uZQAAAAAAAAABAAAAMwAAAAtzc2gtZW
+QyNTUxOQAAACBnO1dnNsxT0TJfP4Jgb9fzBJXRLiWrvIx44cftqs4mLAAAAJjYNRR+2DUU
+fgAAAAtzc2gtZWQyNTUxOQAAACBnO1dnNsxT0TJfP4Jgb9fzBJXRLiWrvIx44cftqs4mLA
+AAAEDKN3GvoFkLLQdFN+Blk3y/+HQ5rvt7/GALRAWofc/LFGc7V2c2zFPRMl8/gmBv1/ME
+ldEuJau8jHjhx+2qziYsAAAAEHJvb3RAY3QtZGlyLWRldjIBAgMEBQ==
+-----END OPENSSH PRIVATE KEY-----
+EOF
+    chmod 600 /var/lib/devhost/ssh_bootstrap_key
+    fi
+
+    case "$action" in
+      help)
+        echo "usage: fc-manage-dev-vms [options] ACTION VM_NAME [HYDRA_EVAL]
+
+Options:
+-m           Memory of the VM. Format: Number + Unit in 1 Letter. Example: 1024M
+-c           CPU Cores of the VM. Format: Number
+
+Arguments:
+ACTION       must be one of help, destroy, ensure
+VM_NAME      the name of the vm to be created or destroyed
+HYDRA_EVAL   ID of the FC NixOS Hydra evaluation. Used for determining channel and base image of the VM. Required when ACTION=ensure
+"
+;;
+      destroy)
+        vm=''${2?need to specify vm name}
+        rm /etc/devhost/vm-configs/$vm.nix
+        rm -r /var/lib/devhost/vms/$vm || true
+        fc-manage -v -b
+      ;;
+      ensure)
+        vm=''${2?need to specify vm name}
+        eval_id=''${3?need to specify Flying Circus Hydra eval id}
+        aliases=''${4}
+        aliases=$(echo $aliases | sed 's| |" "|g;s|.*|"&"|')
+        channel_url="https://hydra.flyingcircus.io/build/$(curl -H "Accept: application/json" -L https://hydra.flyingcircus.io/eval/$eval_id/job/release | jq -r .id)/download/1/nixexprs.tar.xz"
+        channel_path=''${channel_url#file://}
+
+        vm_exists=false
+        if [[ -f "/etc/devhost/vm-configs/$vm.nix" ]]; then
+          vm_exists=true
+        fi
+        mkdir -p /var/lib/devhost/vms/$vm
+        cat > /etc/devhost/vm-configs/$vm.nix <<EOF
+## DO NOT TOUCH!
+## Managed by fc-manage-dev-vms
+{ ... }: {
+  flyingcircus.roles.devhost.virtualMachines = {
+    "$vm" = {
+      memory = "$memory";
+      cores = $cores;
+    };
+  };
+}
+
+EOF
+        if [ "$vm_exists" = false ]; then
+          vm_base_image_store_path=$(curl -H "Accept: application/json" -L https://hydra.flyingcircus.io/eval/$eval_id/job/images.dev-vm | jq -r ".buildproducts[] | select(.subtype==\"img\").path")
+          nix-store -r $vm_base_image_store_path
+          cp $vm_base_image_store_path /var/lib/devhost/vms/$vm/rootfs.qcow2
+          fc-manage -v -b
+          until ping -qc 1 dev-vm
```

Comment:
This has one huge problem: It's not supported that multiple VMs start up at the same time.

Just had one idea to fix this: We use SLAAC based on the MAC Address of the VM in the script. The VM could still have a dynamic v4 that is used for the end user with dns.